src/storage/copier/csv_reader.cpp

```diff
@@ -0,0 +1,515 @@
+    if (mode == ParserMode::PARSING_HEADER) {
+        return true;
+    }
+    linenr++;
+    if (row_empty) {
+        row_empty = false;
+        if (return_types.size() != 1) {
+            if (mode == ParserMode::PARSING) {
+                // Set This position to be null
+                ;
+            }
+            column = 0;
+            return false;
+        }
+    }
+    if (column < return_types.size()) {
+        throw CopyException(StringUtils::string_format(
+            "Error in file {} on line {}: expected {} values per row, but got {}", filePath, linenr,
+            return_types.size(), column));
+    }
+    if (mode == ParserMode::PARSING &&
+        columnSizes[0] >= common::StorageConstants::NODE_GROUP_SIZE) {
+        Flush(insert_chunk, buffer_idx);
+        return true;
+    }
+
+    column = 0;
+    return false;
+}
+
+bool BaseCSVReader::Flush(DataChunk& insert_chunk, uint64_t buffer_idx, bool try_add_line) {
+    if (parse_chunk.getNumValueVectors() == 0) {
+        return true;
+    }
+    for (uint64_t c = 0; c < parse_chunk.getNumValueVectors(); c++) {
+        auto parse_vector = parse_chunk.getValueVector(c);
+        auto result_vector = insert_chunk.getValueVector(c);
+        insert_chunk.getValueVector(c)->state->selVector->selectedSize =
+            parse_chunk.getValueVector(c)->state->selVector->selectedSize;
+        insert_chunk.insert(c, parse_chunk.getValueVector(c));
+    }
+    InitParseChunk(parse_chunk.getNumValueVectors());
+    return true;
+}
+
+void BaseCSVReader::InitParseChunk(column_id_t numCols) {
+    parse_chunk = DataChunk(numCols);
+}
+
+BufferedCSVReader::BufferedCSVReader(const std::string& filePath,
+    common::CSVReaderConfig* csvReaderConfig, catalog::TableSchema* tableSchema,
+    storage::MemoryManager* memoryManager = nullptr)
+    : BaseCSVReader(filePath, csvReaderConfig, tableSchema), bufferSize(0), position(0), start(0) {
+    Initialize(tableSchema->getProperties(), memoryManager);
+}
+
+void BufferedCSVReader::Initialize(
+    std::vector<kuzu::catalog::Property*> properties, storage::MemoryManager* memoryManager) {
+    // PrepareComplexParser();
+    for (auto property : properties) {
+        return_types.push_back(*property->getDataType());
+    }
+    fd = open(filePath.c_str(), O_RDONLY);
+    ResetBuffer();
+    ReadHeader(csvReaderConfig->hasHeader);
+    // Initializing parse_chunk DataChunk
+    parse_chunk = DataChunk(properties.size());
+    for (int i = 0; i < properties.size(); i++) {
+        auto type = properties[i]->getDataType();
+        auto typeID = type->getLogicalTypeID();
+        auto v = std::make_shared<ValueVector>(*type, memoryManager);
+        parse_chunk.insert(i, v);
+    }
+    mode = ParserMode::PARSING;
+}
+
+void BufferedCSVReader::ResetBuffer() {
+    buffer.reset();
+    bufferSize = 0;
+    position = 0;
+    start = 0;
+    cachedBuffers.clear();
+}
+
+void BufferedCSVReader::ReadHeader(bool hasHeader) {
```

Comment:
This function is a bit confusing, as it is named as `ReadHeader`, and from the function definition, I can tell parsing header is the main stuff, but still this function takes in a parameter to allow skiping reading header. And regardless of has header or not, the function will set `mode` to `PARSING`.
It seems the caller should first switch on `hasHeader` to decide to call this function or not, which is more clear.